### DISCUSSION TITLE
posts/elf-standard/

### DISCUSSION THREAD

**Discussion by giscus** [brant-ruan/fernweh-comments#5] [Announcements]
# posts/elf-standard/

前言 译者注：
可移植格式规格说明，版本 1.1。 来自工具接口标准 (TIS)。 已有 TIS 1.2版，但ELF文件格式这部分与1.1版本差别不大。 ELF:可重定位/链接格式
ELF文件格式作为应用程序二进制接口(ABI)，最初由Unix系统实验室(USL)编写并发布。工具接口标准委员会(TIS)选择改进后的ELF标准作为在32位Intel架构上的可移植目标文件格式，适用于多种操作系统。
ELF标准使软件开发变得简单，它为开发者提供了跨多平台的二进制接口定义。这将减少不同接口实现的数目，从而减少重新编码和重新编译的需要。
关于本文档
本文档的预期读者是为多种32位系统环境开发目标文件或可执行文件的开发人员。
文档分为以下几个部分：
“目标文件”描述了3种主要的ELF目标文件格式[1]。 “程序装载和动态链接”描述了在创建运行时程序时目标文件信息和操作系统的执行过程。 “C库”列举了libsys(即标准C(ANSI C)和libc库实例)包含的符号，以及libc库实例需要的全局数据符号。 注：X86架构参考文档变为Intel架构。
1 目标文件 介绍 第一部分描述了ABI目标文件格式——ELF。它主要有三种目标文件类型：
可重定位文件包含了用来和其他目标文件链接起来生成可执行文件或共享目标文件的代码和数据。 可执行文件包含一个可执行的程序。它详定了exec(BA_OS)该怎样创建程序的进程镜像。 共享目标文件包含了在两种情境下用于链接的代码和数据。第一，链接[参照 ld(SD_CMD)]可能将它和其他可重定位文件以及共享目标文件合起来生成另一个目标文件；第二，动态链接器把它和一个可执行文件以及其他共享目标文件结合，创造一个进程镜像。 目标文件由汇编器和链接器创建，是可以在处理器上直接运行的二进制程序。那些需要虚拟机才能够执行的程序，如shell脚本，不属于这一范围。
在介绍性的材料之后，第一部分聚焦于文件格式以及它和构建程序的关系。第二部分也描述了部分目标文件，该部分聚焦于执行程序所需的必要信息。
文件格式 目标文件参与程序链接(构建程序)和程序执行(运行程序)的过程。考虑到方便性和效率，目标文件格式提供了并行的多种视角来描述文件内容，反映出了不同程序活动的需要。图1-1展示了一个目标文件的组织结构。
图1-1：目标文件格式
- 链接视角 - ELF头 - 程序头表(可选) - 第1节(section) - ... - 第n节 - ... - ... - 节头表 - 执行视角 - ELF头 - 程序头表 - 第1段(segment) - 第2段 - ... - 节头表(可选) ELF头存在于目标文件开头，相当于描述文件组织结构的“路线图”。在链接视角中，节构成了目标文件信息主体：指令，数据，符号表和重定位信息等等。第一部分后面将对特殊节进行描述。第二部分讨论段以及文件的程序执行视角。
程序头表(如果有)告诉系统怎样创造进程镜像。用于构建进程镜像(执行程序)的文件必须有一个程序头表；可重定位文件不需要有。
节头表包含了描述文件节的信息。每一个节在表中都占有一项；每一项都给出了如节名，节大小等信息。链接期间使用的文件必须有一个节头表；其他目标文件则可有可没有。
注：虽然图1-1中程序头表紧跟在ELF头后面，节头表跟在节后面，但实际文件可能有所差异。另外，节和段没有限定顺序。只有ELF头的位置是固定的。
数据表示 目标文件格式支持多种符合8比特一个字节和32位架构标准的处理器。然而，它可以扩展到更大(或更小)的架构上。因此，目标文件使用独立于机器的格式来表示一些控制数据，这允许以统一的方式识别目标文件并解释它们的内容。目标文件中的剩余数据使用目标处理器上的编码方式，与创建文件的机器无关。

https://blog.wohin.me/posts/elf-standard/

**Comment by brant-ruan** [comment]
想起2016年的暑假，读了《约翰·克里斯朵夫》，翻译了ELF文件标准。
